src/common/row_operations/row_gather.cpp

```diff
@@ -0,0 +1,98 @@
+//===--------------------------------------------------------------------===//
+// row_gather.cpp
+// Description: This file contains the implementation of the gather operators
+//===--------------------------------------------------------------------===//
+
+#include "duckdb/common/exception.hpp"
+#include "duckdb/common/operator/constant_operators.hpp"
+#include "duckdb/common/row_operations/row_operations.hpp"
+#include "duckdb/common/types/row_layout.hpp"
+
+namespace duckdb {
+
+using ValidityBytes = RowLayout::ValidityBytes;
+
+template <class T>
+static void TemplatedGatherLoop(Vector &rows, const SelectionVector &row_sel, Vector &col,
+                                const SelectionVector &col_sel, idx_t count, idx_t col_offset, idx_t col_no) {
+	// Precompute mask indexes
+	idx_t entry_idx;
+	idx_t idx_in_entry;
+	ValidityBytes::GetEntryIndex(col_no, entry_idx, idx_in_entry);
+
+	auto ptrs = FlatVector::GetData<data_ptr_t>(rows);
+	auto data = FlatVector::GetData<T>(col);
+	auto &col_mask = FlatVector::Validity(col);
+
+	for (idx_t i = 0; i < count; i++) {
+		auto row_idx = row_sel.get_index(i);
+		auto row = ptrs[row_idx];
+		auto col_idx = col_sel.get_index(i);
+		data[col_idx] = Load<T>(row + col_offset);
+		ValidityBytes row_mask(row);
+		if (!row_mask.RowIsValid(row_mask.GetValidityEntry(entry_idx), idx_in_entry)) {
```

Comment:
For the future: this check could be optimized away if we know the column does not contain null values (e.g. through statistics).

Reply:
Yeah, I noticed that the one of the scatter code implementations was not checking whether the block was all valid, and the shared implementation does this, so there have been some wins.